configure.ac

```diff
@@ -1299,14 +1299,22 @@ AC_MSG_NOTICE([GnuTLS library support: ${with_gnutls:=auto} ${LIBGNUTLS_PATH} ${
 dnl User may specify OpenSSL is needed from a non-standard location
 SQUID_OPTIONAL_LIB(openssl,[OpenSSL],[LIBOPENSSL])
 AH_TEMPLATE(USE_OPENSSL,[OpenSSL support is available])
-## OpenSSL is default disable due to licensing issues on some OS
+AS_IF([test "x$with_openssl" != "xno"],[
+  SQUID_STATE_SAVE(squid_openssl3_state)
+  CPPFLAGS="$LIBOPENSSL_CFLAGS $CPPFLAGS"
+  LIBS="$LIBS $LIBOPENSSL_PATH"
+  PKG_CHECK_MODULES([LIBOPENSSL],[openssl >= 3],[with_openssl="yes"],[:])
+  SQUID_STATE_ROLLBACK(squid_openssl3_state)
+])
+## OpenSSL < 3 is default disable due to licensing issues on some OS
 if test "x$with_openssl" = "xyes"; then
   CPPFLAGS="$LIBOPENSSL_CFLAGS $CPPFLAGS"
   AC_CHECK_HEADERS( \
     openssl/asn1.h \
```

Comment:
My understanding is that it is just a bit of juggling to simulate --with-openssl, results in with_openssl="yes" as if --with-openssl had been specified on the command line. No change wrt AC_CHECK_HEADERS and other actions.

But I am a bit confused by the LIBS&CFLAGS juggling, and how it works with a non-standard path --with-openssl=/path/to/openssl-install but original code is pretty much as confusing.

And also the PKG_CHECK_MODULES is invoked twice, which should be harmless but not sure it's intentional.

But is this at all working? The SQUID_OPTIONAL_LIB() already sets with_openssl="no" for the default case I think

  AS_IF([test "x$withval" = "x"],[m4_translit([with_$1], [-+.], [___])=no])

Reply:
> My understanding is that it is just a bit of juggling to simulate --with-openssl, results in with_openssl="yes" as if --with-openssl had been specified on the command line. No change wrt AC_CHECK_HEADERS and other actions.

I agree. I misread the diff because the code essentially overwrites user configuration. Sorry. I consider my change request resolved. @yadij, I will keep it open because Henrik's comment discusses other issues, but please feel free to close this thread.

I also agree that there are other problems here. I did not want to dive into them until my higher-level concern was addressed. I recommend opening dedicated change request(s) for those concerns rather than piggybacking them to this addressed change request. For example, The misnamed SQUID_OPTIONAL_LIB() is documented to target default-disabled libraries. If the documented "Default-enable OpenSSL if pkg-config finds OpenSSL v3+." goal of this PR remains unchanged, then OpenSSL should no longer be considered a "default-disabled" library: Since v3+ is the "stable present and future", we should make our primary decisions based on that, and then take care of any "temporary exceptions" like v1-.

The original code is terrible. This PR does not have to improve it, but we should not make things worse, of course.

Reply:
I'd suggest dropping this fragment, and work on openssl3 enabled by default in another follow-up change after this PR. It's unrelated to adding support.

Reply:
Dropped in[0f14f83](https://github.com/squid-cache/squid/pull/694/commits/0f14f83d5e83e524291f4ba8ea2fac83e84722a3)